File: Codewars/3kyu/3kyu_can_you_get_the_loop/src/main.py
class Node(object):
    pass

def loop_size(node):
    nodes = []
    tail_node = None

    while(not node in nodes):
        nodes.append(node)
        node = node.next

    return len(nodes) - nodes.index(node)
   
node1 = Node()
node1.next = node1

print loop_size(node1) # 1

node1 = Node()
node2 = Node()

node1.next = node2
node2.next = node1

print loop_size(node1) # 2

node1 = Node()
node2 = Node()
node3 = Node()

node1.next = node2
node2.next = node3
node3.next = node2

print loop_size(node1) # 2

# Make a short chain with a loop of 3
node1 = Node()
node2 = Node()
node3 = Node()
node4 = Node()
node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node2

print loop_size(node1) # 3

nodes = [Node() for _ in xrange(50)]
for node, next_node in zip(nodes, nodes[1:]):
    node.next = next_node
nodes[49].next = nodes[21]

print loop_size(nodes[0]) # 29

nodes = [Node() for _ in xrange(3904)]
for node, next_node in zip(nodes, nodes[1:]):
    node.next = next_node
nodes[3903].next = nodes[2817]

print loop_size(nodes[0]) # 1087

File: Codewars/3kyu/3kyu_can_you_get_the_loop/src/main.js
class Node {}

function loop_size(node) {
    let rabbit = node.next.next;
    let tottle = node.next;

    while(rabbit != tottle) {
        rabbit = rabbit.next.next;
        tottle = tottle.next;
    }

    let count = 1;
    rabbit = rabbit.next;
    while(tottle != rabbit) {
        count++;
        rabbit = rabbit.next;
    }
    return count;
}

let node1 = new Node();
node1.next = node1;

console.log(loop_size(node1)); // 1

node1 = new Node();
let node2 = new Node();

node1.next = node2
node2.next = node1

console.log(loop_size(node1)); // 2

node1 = new Node();
node2 = new Node();
let node3 = new Node();

node1.next = node2
node2.next = node3
node3.next = node2

console.log(loop_size(node3)); // 2